Add optional `outputSize` to `opts` of `webContents.capturePage([rect, opts])`

Imperat opened the pull request:
#### Description of Change
This PR adds optional parameter `outputSize` to `opts` of `webContents.capturePage([rect, opts])`. The main purpose of this parameter is to improve speed and resources consumption when generating small thumbnails of web pages.

<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Contributors guide: https://github.com/electron/electron/blob/main/CONTRIBUTING.md
-->

#### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] PR description included and stakeholders cc'd
- [ ] `npm test` passes
- [ ] tests are [changed or added](https://github.com/electron/electron/blob/main/docs/development/testing.md)
- [x] relevant documentation, tutorials, templates and examples are changed or added
- [x] [PR release notes](https://github.com/electron/clerk/blob/main/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/main/README.md#examples).

#### Release Notes

notes: Added optional `outputSize` to `opts` of `webContents.capturePage([rect, opts])`.

Imperat commented:
cc @ckerr @codebytere @zcbenz

codebytere commented on docs/api/browser-window.md:
```diff
@@ -1162,6 +1162,9 @@ Returns `boolean` - Whether the window's document has been edited.
 * `opts` Object (optional)
   * `stayHidden` boolean (optional) -  Keep the page hidden instead of visible. Default is `false`.
   * `stayAwake` boolean (optional) -  Keep the system awake instead of allowing it to sleep. Default is `false`.
+  * `outputSize` Object (optional)
```
We should specify the units being used for both width and height.

Imperat replied:
added: _The desired width and height in pixels (positive numbers) of the returned image_

codebytere commented on docs/api/web-contents.md:
```diff
@@ -1519,12 +1519,16 @@ console.log(requestId)
 * `opts` Object (optional)
   * `stayHidden` boolean (optional) -  Keep the page hidden instead of visible. Default is `false`.
   * `stayAwake` boolean (optional) -  Keep the system awake instead of allowing it to sleep. Default is `false`.
+  * `outputSize` Object (optional)
```
Same as above - units need specification!

Imperat replied:
added: _The desired width and height in pixels (positive numbers) of the returned image_

codebytere commented on shell/browser/api/electron_api_web_contents.cc:
```diff
@@ -3533,16 +3535,17 @@ v8::Local<v8::Promise> WebContents::CapturePage(gin::Arguments* args) {
-  if (scale > 1.0f)
-    bitmap_size = gfx::ScaleToCeiledSize(view_size, scale);
+  if (bitmap_size.IsEmpty()) {
+    // By default, the requested bitmap size is the view size in screen
```
This should also perhaps be specified in the documentation - this might not be assumed behavior by users. Beyond that - is this something we want to do by default? Or is this something that we might want to delegate to users as an option when passing `width` and `height`?

Imperat replied:
I've made changes in documentation. 

>Beyond that - is this something we want to do by default?
Not sure, I've made changes in that way to not change or affect existing apps which rely on this API. Currently, Electron does it by default.

> Or is this something that we might want to delegate to users as an option when passing width and height
For the sake of practicality I do not want to add another option to this small PR to control enablement / exposure of scaling feature. However, I am happy to raise a separate PR to introduce this option once the current PR reaches `main` :)

codebytere replied:
Fair enough! I don't feel strongly about it really :)

codebytere reviewed (changes requested):
Thanks for adding this! I think the performance implications here especially for lower-bandwidth users add a compelling use case for this. We'll definitely also want to add some tests here to verify the new behavior.

ckerr reviewed (commented):
Code LGTM.

+1 on @codebytere's docs change suggestions

Imperat commented:
Hi @ckerr @codebytere 
Thanks for reviewing this PR 🙏 
I've added two tests to `spec/api-browser-view-spec.ts` and `spec/api-browser-window-spec.ts`. I was unable to run all of the tests via `npm test` (it takes some time), but I've hardcoded `spec-runner.ts` and verified that `BrowserView.capturePage(rect)` and `BrowserWindow.capturePage(rect)` passed.

I've also updated documentation as you requested 👍 

Please have a look

codebytere commented on docs/api/web-contents.md:
```diff
@@ -1519,13 +1519,19 @@ console.log(requestId)
 If you would like the page to stay hidden, you should ensure that `stayHidden` is set to true.
 
+The size of returned [NativeImage](native-image.md) will be `outputSize` if it was provided.
+Otherwise, the size would be large enought to fit all of the available pixels in the current system.
```
We should validate this with a test! Ex. validate that if the default size would be `{ 200, 200 }` and i set outputSize to `{ 1500, 150 }` the resultant image should have a size of `{ 1500, 150 }`.

Imperat replied:
Good suggestion!
I've elaborated the test into three scenarios: 
1. requested size is smaller than captured area
2. one dimension of requested size is bigger than captured area
3. both dimensions are bigger than captured area

codebytere commented on docs/api/browser-window.md:
```diff
@@ -1162,6 +1162,18 @@ Returns `boolean` - Whether the window's document has been edited.
   * `stayAwake` boolean (optional) -  Keep the system awake instead of allowing it to sleep. Default is `false`.
+  * `outputSize` [Size](structures/size.md) (optional) - The desired width and height in pixels (positive numbers) of the returned image.
+
+Returns `Promise<NativeImage>` - Resolves with a [NativeImage](native-image.md)
+
+Captures a snapshot of the page within `rect`. Omitting `rect` will capture the whole visible page.
+The page is considered visible when its browser window is hidden and the capturer count is non-zero.
+If you would like the page to stay hidden, you should ensure that `stayHidden` is set to true.
```
Looks like this got duplicated? It's [here](https://github.com/electron/electron/pull/41281/files#diff-e8a455d5c2bb8c09554011e09a285d6e66c3c681082ae9998f2a4a026eaf0b03R1178) already.

Imperat replied:
Sorry, I missed that. [Fixed in the last commit](https://github.com/electron/electron/pull/41281/commits/861d79b9f12cb9b09055d2580925203110e54bb1)

codebytere reviewed (commented):
Thanks for addressing this - once there's one more test for large edge size validation and the duplicated docs get removed i think we can get this moving!

codebytere reviewed (approved):
API LGTM

jkleinsc commented on docs/api/web-contents.md:
```diff
@@ -1519,13 +1519,19 @@ console.log(requestId)
 If you would like the page to stay hidden, you should ensure that `stayHidden` is set to true.
 
+The size of returned [NativeImage](native-image.md) will be `outputSize` if it was provided.
+Otherwise, the size would be large enought to fit all of the available pixels in the current system.
```
Same comment above. - this statement is confusing as is.

jkleinsc reviewed (changes requested):
The documentation needs clarity added around the description of what happens when outputSize is not provided.

nornagon commented on docs/api/browser-window.md:
```diff
@@ -1154,10 +1154,18 @@ Returns `boolean` - Whether the window's document has been edited.
+If you would like the page to stay hidden, you should ensure that `stayHidden` is set to true.
+
+The size of returned [NativeImage](native-image.md) will be `outputSize` if it was provided.
+Otherwise, the size would be large enought to fit all of the available pixels in the current system.
```
This is confusing, what does "all of the available pixels in the current system" mean?

How does this behave if the aspect ratio of `outputSize` differs from the aspect ratio of `rect`?

Imperat replied:
For "all of the available pixels":
Initially, I took this phrase somewhere from Chromium sources, but I agree it sounds confusing.

I rephrased it to state the default value instead:

> The desired width and height in pixels (positive numbers) of the returned image. Default is the size of WebContents in physical pixels, that is WebContents height and WebContents width multiplied by device pixel ratio.

For aspect ratio:

I've added a note:
> **Note:** When using `outputSize` it is responsibility of caller to preserve aspect ratio. If aspect ratio of `outputSize` is not the same as page aspect ratio the output image will be scaled.

Imperat commented:
Hi @jkleinsc @nornagon 

Thank you for your feedback. I apologize for not being able to update the PR earlier when I received your review.
I made some changes now. Could you please have a second look? 🙏

samuelmaddock commented on spec/api-browser-window-spec.ts:
```diff
@@ -2307,6 +2307,76 @@ describe('BrowserWindow module', () => {
+        });
+      });
+
+      describe('when both dimensions of `outputSize` are bigger than both dimensions of the captured area', () => {
+        it('returns image with requested size', async () => {
```
It's fine as is, but this should probably just use one `it` like
```suggestion
      it('returns image with requested size when both dimensions of `outputSize` are bigger than both dimensions of the captured area', async () => {
```

samuelmaddock reviewed (approved):
API LGTM

samuelmaddock commented:
@Imperat your PR might need to be rebased to resolve the build issues being seen.

samuelmaddock commented:
I'm adding a wip label on this PR until the requested changes can be implemented. Please tag `@electron/wg-api` once it's ready to be reviewed again and we'll take another look! 🙇

codebytere commented on spec/api-browser-window-spec.ts:
```diff
@@ -2434,6 +2434,70 @@ describe('BrowserWindow module', () => {
+        });
+
+        expect(image.getSize().width).to.be.lessThan(1500);
+        expect(image.getSize().height).to.be.lessThan(1500);
```
```suggestion
        expect(image.getSize()).to.deep.equal({ width: 200, height: 200 });
```

Imperat commented:
Hi @jkleinsc 

Thank you for your effort to attempt to merge this PR 🙏 
I will try to return to it once I have some free time
